src/coreclr/jit/codegenxarch.cpp

```diff
@@ -6308,6 +6308,31 @@ void CodeGen::genIntToIntCast(GenTreeCast* cast)
         genIntCastOverflowCheck(cast, desc, srcReg);
     }
 
+    // Code size optimization: use "cdqe" and "cdwe" instead
+    // of "movsxd rax, eax" and "movsx eax, ax", respectively.
+    if ((srcReg == REG_EAX) && (srcReg == dstReg))
+    {
+        emitAttr attribute = EA_UNKNOWN;
+#ifdef TARGET_64BIT
+        if (desc.ExtendKind() == GenIntCastDesc::SIGN_EXTEND_INT)
+        {
+            attribute = EA_8BYTE;
+        }
+#endif
+        if ((desc.ExtendKind() == GenIntCastDesc::SIGN_EXTEND_SMALL_INT) &&
+            (desc.ExtendSrcSize() == sizeof(int16_t)))
```

Comment:
It is notable that the old code did `movsx rax, ax` in this case, i. e. extended to the full register width, while the new code only extends to 32 bits. It ought not to have ill effects as the node type was and is `int`, so its user can only read the lower 32 bits of the register.

However, the lack of representation for `CAST(long <- byte <- int)` does make optimizing cases like `*longPtr = *bytePtr` a bit challenging. Today (and in this PR) we end up with a redundant sign extension:
```asm
       movsx    rax, byte  ptr [rdx]
       cdqe     
       mov      qword ptr [rcx], rax
```

Reply:
We are handling redundant sign extensions already here: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/emitxarch.cpp#L4365

This PR should likely extend that to include `cdq`/`cdqe` (for zx/sx, respectively)

Reply:
Well, that method does not handle it today - we end up with `movsx`/`movsxd` pair.

One thing I wanted to avoid is plumbing `cdqe` through as a "`mov` instruction", because, while it conceptually is, I do not know how many method assume things like "if it is a `mov`, it must have two operands" and similar (it also seemed not worthwhile - there can just be a special case for it in the "is redundant mov" optimization once that's enhanced to handle this deficiency).

Reply:
I think that would be fixed by keeping it as `movsx eax, ax` and simply special casing that specific setup during size computation and actual assembly emit. It would also avoid needing to special case other code paths (not every sign extension comes from `genIntToIntCast` iirc) and avoid needing to special case an unknown number of other paths to also handle `cdqe`. Instead, its just 2 well known locations which explicitly handle the size computation and byte emission for `ins reg, reg`

Reply:
> I think that would be fixed by keeping it as `movsx eax, ax`

I see your point. One downside is that it would be a little worse TP-wise since we'd be "cold" logic to a general path, also used in MinOpts. Presumably it's not going to cost us much.

Reply:
> One downside is that it would be a little worse TP-wise since we'd be adding "cold" logic to a general path, also used in MinOpts. Presumably it's not going to cost us much.

Agreed. We do enough other checks for similar scenarios I don't think this will matter much in practice.

If it does, the bigger emitter wins are likely elsewhere and would come out of a larger refactoring to avoid duplicating all the checks and prefix additions (which we do multiple times per instruction in many cases, even in release builds).